filebeat/beater/channels.go

```diff
@@ -0,0 +1,103 @@
+package beater
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/filebeat/registrar"
+	"github.com/elastic/beats/filebeat/spooler"
+)
+
+type spoolerOutlet struct {
+	wg      *sync.WaitGroup
+	done    <-chan struct{}
+	spooler *spooler.Spooler
+}
+
+type publisherChannel struct {
+	done chan struct{}
+	ch   chan []*input.Event
+}
+
+type registrarLogger struct {
+	done      chan struct{}
+	registrar *registrar.Registrar
+}
+
+type finishedLogger struct {
+	wg *sync.WaitGroup
+}
+
+func newSpoolerOutlet(
+	done <-chan struct{},
+	s *spooler.Spooler,
+	wg *sync.WaitGroup,
+) *spoolerOutlet {
+	return &spoolerOutlet{
+		done:    done,
+		spooler: s,
+		wg:      wg,
+	}
+}
+
+func (o *spoolerOutlet) OnEvent(event *input.Event) bool {
+	if o.wg != nil {
+		o.wg.Add(1)
+	}
+
+	select {
+	case <-o.done:
+		if o.wg != nil {
+			o.wg.Done()
+		}
+		return false
+	case o.spooler.Channel <- event:
+		return true
+	}
+}
+
+func newPublisherChannel() *publisherChannel {
+	return &publisherChannel{
+		done: make(chan struct{}),
+		ch:   make(chan []*input.Event, 1),
+	}
+}
+
+func (c *publisherChannel) Close() { close(c.done) }
+func (c *publisherChannel) Send(events []*input.Event) bool {
+	select {
+	case <-c.done:
+		return false
+	case c.ch <- events:
+		return true
+	}
+}
+
+func newRegistrarLogger(reg *registrar.Registrar) *registrarLogger {
```

Comment:
Idea is to hook up all the components in beater Run method. The registrarLogger is used to connect the publisher and the registrar.

One of the reasons it's not in registrar: due to all components being async workers (which becomes super aparent in provided interfaces), we have to be able to close the channel without stopping the registrar. This is achieved via registrarLogger being closed right before `publisher.Stop` is executed.